src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs

```diff
@@ -0,0 +1,1161 @@
+
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics;
+
+public class InterceptorsTests : CSharpTestBase
+{
+    private static readonly (string, string) s_attributesSource = ("""
+        namespace System.Runtime.CompilerServices;
+
+        [AttributeUsage(AttributeTargets.Method)]
+        public sealed class InterceptableAttribute : Attribute { }
+
+        [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+        public sealed class InterceptsLocationAttribute : Attribute
+        {
+            public InterceptsLocationAttribute(string filePath, int line, int column)
+            {
+            }
+        }
+        """, "attributes.cs");
+
+    [Fact]
+    public void IsInterceptable()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            class C
+            {
+                [Interceptable]
+                public static void InterceptableMethod() { Console.Write("interceptable"); }
+                
+                public static void NotInterceptable() { Console.Write("not interceptable"); }
+            }
+            """;
+
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, symbolValidator: verify, sourceSymbolValidator: verify);
+        verifier.VerifyDiagnostics();
+
+        void verify(ModuleSymbol module)
+        {
+            var method = module.GlobalNamespace.GetMember<MethodSymbol>("C.InterceptableMethod");
+            Assert.True(method.IsInterceptable);
+            Assert.Equal(MethodKind.Ordinary, method.MethodKind);
+
+            method = module.GlobalNamespace.GetMember<MethodSymbol>("C.NotInterceptable");
+            Assert.False(method.IsInterceptable);
+            Assert.Equal(MethodKind.Ordinary, method.MethodKind);
+        }
+    }
+
+    [Fact]
+    public void StaticInterceptable_StaticInterceptor_NoParameters()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            class C
+            {
+                [Interceptable]
+                public static void InterceptableMethod() { Console.Write("interceptable"); }
+
+                public static void Main()
+                {
+                    InterceptableMethod();
+                }
+            }
+
+            class D
+            {
+                [InterceptsLocation("Program.cs", 11, 9)]
+                public static void Interceptor1() { Console.Write("interceptor 1"); }
+            }
+            """;
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, expectedOutput: "interceptor 1");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact(Skip = "PROTOTYPE(ic): produce an error here")]
```

Comment:
>produce an error here [](http://example.com/codeflow?start=33&length=21)

Should this be an error because the intercepting method is private? If so, is intercepting with a private method problematic?

Reply:
Yeah, an error occurs at runtime because the interceptor is not accessible from the call site.